### ISSUE TITLE
Problematic celestial body colliders

### ISSUE THREAD

**Issue by WKFO** [pioneerspacesim/pioneer#4402]
Long-term issue (?)

### Observed behaviour
Planetary/stellar body colliders don't make any sense. Sometimes they don't even work.
Extreme example: https://youtu.be/UTMK-oK3xZs

### Expected behaviour
https://youtu.be/yUJjQWsStV8

### Steps to reproduce
Crash into a planet, not head-first but like landing an airplane.

__My pioneer version (and OS):__
https://github.com/pioneerspacesim/pioneer/commit/12cc4d80d786100fe5d3ae709d938315dad452a0, Windows

**Comment by fluffyfreak** [comment]
Just for this, I'm gonna make stars properly kill you :P

**Comment by fluffyfreak** [comment]
@WKFO hi, I've been trying to reproduce this but I just can't seem to manage it, can you still get it to happen? If so where and have you got any tips for how to do it?

**Comment by WKFO** [comment]
@fluffyfreak I prepared a save file for you: https://www.dropbox.com/s/ow953dei4trvq2k/Ethandce2?dl=0
If you don't touch the controls, the Sinonatrix will pass right through the star.

It used to happen more on terrestrial planets but I think it doesn't happen anymore. Either a helpful bunch of gamma rays struck the github servers and fixed it, or it might have something to do with https://github.com/pioneerspacesim/pioneer/pull/4588 though I am not sure.

**Comment by impaktor** [comment]
@WKFO just for completeness, which version of pioneer are you on? e.g. build date, or hash, as shown in bottom right corner of main menu screen.

**Comment by WKFO** [comment]
https://github.com/pioneerspacesim/pioneer/commit/12cc4d80d786100fe5d3ae709d938315dad452a0

**Comment by mike-f1** [comment]
I think #4588 is fixing this thing, though with its own limits and with following considerations.

`Star` inherit from `TerrainBody` which calls below function which in turn calls `Terrain` fractals for height, through `GeoSphere` (which inherit from `BaseSphere`):
https://github.com/pioneerspacesim/pioneer/blob/b2258e261bb47fbb30908dbe018fda1134651ea3/src/TerrainBody.cpp#L138

But a odd thing is here:
https://github.com/pioneerspacesim/pioneer/blob/b2258e261bb47fbb30908dbe018fda1134651ea3/src/TerrainBody.cpp#L29-L41

...as it seems any type of star require a `GeoSphere`, as opposed as a gas giants, which have their own "Sphere", which is calling:

https://github.com/pioneerspacesim/pioneer/blob/b2258e261bb47fbb30908dbe018fda1134651ea3/src/GasGiant.h#L43

showing that GasGiant terrain is "flat"/"spherical" and thus it has constant heigth (no fractal for
height here...)

Anyway it is also showing an older [issue](https://github.com/pioneerspacesim/pioneer/issues/1834) here:
when you (accidentally :D ) go under a star surface there's a lot of triangles!

To me, seems that a specialization for Stars as
`StarSphere: public BaseSphere` which limit 
1) depth of patches (...I recently discovered that Patches are 
16k(IIRC?) points, and there are 16 levels of 4 patch each)?
2) and re-inherit `GetHeight` as in `GasGiant`

may improve _a lot_ render and give better collision results when people try to "go submarin-ing in a Black hole sun" :D

Some little question @WKFO :
1) Temperature and Pressure: why your ship don't explode or implode before hitting the surfaces? (...and it could be a better and reasonable behaviour, isn't?)
2) have you tried with time accel? (...I'm asking because #4588 have their own limits...)
3) Could you try doing the same but with (...or "in") a gas giant?

**Comment by fluffyfreak** [comment]
In this case I think, but can't quite prove it yet, that you might not be able to collide with the star because it doesn't think you're within it's frame / radius yet.

Not 100% sure on it, but something wierd is going on.

**Comment by fluffyfreak** [comment]
That is super bizarre. it bails out of the [space.cpp:ln967](https://github.com/pioneerspacesim/pioneer/blob/master/src/Space.cpp#L967) because it doesn't have an `m_astroBody`, it's `nullptr`.
It looks like that should not happen, then suddenly it gets one from somewhere, and it's the "correct" one but not until you're well below the surface of the star... instant death

**Comment by mike-f1** [comment]
https://github.com/pioneerspacesim/pioneer/blob/75e66ce5687ff15e0b90dc5f199ea68b16897a50/src/Space.cpp#L1053

before & after?

**Comment by fluffyfreak** [comment]
Think I've figured it out, going to have to call it a night but this is what appears to be happening.

There's 3 stars in this system: A, B & C. A&B are within 2AU of each other around a grav point.

This is the frame dump:

```C
> Frame structure for 'Ethandce':
>   000001ADB93267B0 (System) distance 0.000000 radius 340282346638528859811704183484516925440.000000
>     000001ADB9327C40 (Ethandce A,B) parent 000001ADB93267B0 (System) distance 748521816113.639160 radius 268307805776.141510
>       000001ADB93261D0 (Ethandce A) parent 000001ADB9327C40 (Ethandce A,B) body 000001ADC92C4980 (Ethandce A) bodyFor 000001ADC92C4980 (Ethandce A) distance 23883755930.083416 radius 26485946618.401184
>       000001ADC92D04C0 (Ethandce B) parent 000001ADB9327C40 (Ethandce A,B) body 000001ADC92C3100 (Ethandce B) bodyFor 000001ADC92C3100 (Ethandce B) distance 218796462421.602264 radius 26485946618.401184
>       000001ADC92CF610 (camera) parent 000001ADB9327C40 (Ethandce A,B) distance 103764474994.870819 radius 0.000000 [rotating]
>       000001ADC92CEA50 (camera) parent 000001ADB9327C40 (Ethandce A,B) distance 48253969576.679001 radius 0.000000 [rotating]
>       000001ADC92CF320 (camera) parent 000001ADB9327C40 (Ethandce A,B) distance 48850574318.132393 radius 0.000000 [rotating]
>     000001ADC92CED40 (Ethandce C) parent 000001ADB93267B0 (System) body 000001ADC92C4B40 (Ethandce C) bodyFor 000001ADC92C4B40 (Ethandce C) distance 9219086647226.056641 radius 462368399999.870483
>       000001ADC92CD8B0 (Ethandce C a) parent 000001ADC92CED40 (Ethandce C) body 000001ADA643FBE0 (Ethandce C a) bodyFor 000001ADA643FBE0 (Ethandce C a) distance 192222305814.079407 radius 352432637.827639
>         000001ADC92CF900 (Ethandce C a) parent 000001ADC92CD8B0 (Ethandce C a) body 000001ADA643FBE0 (Ethandce C a) bodyFor 000001ADA643FBE0 (Ethandce C a) distance 0.000000 radius 89814083.177140 [rotating]
>       000001ADC92CDBA0 (Ethandce C b) parent 000001ADC92CED40 (Ethandce C) body 000001ADA643FE80 (Ethandce C b) bodyFor 000001ADA643FE80 (Ethandce C b) distance 377401408321.237671 radius 354844481.351719
>         000001ADC92CFEE0 (Ethandce C b) parent 000001ADC92CDBA0 (Ethandce C b) body 000001ADA643FE80 (Ethandce C b) bodyFor 000001ADA643FE80 (Ethandce C b) distance 0.000000 radius 90195992.619177 [rotating]
```

Now I think what's happening is that the ship is within the parent frame, which does not have an `m_astroBody` because it's just a grav-point. Whereas it should really be within the frame of one of the children, specifically `Ethandce A` which is the one it should have hit and eventually does hit if you slow down and dive towards it's centre.

Will try to look into it more tomorrow but pinging @jaj22 for bright ideas.

**Comment by WKFO** [comment]
> Some little question @WKFO 👍 
> 
>     1. Temperature and Pressure: why your ship don't explode or implode before hitting the surfaces? (...and it could be a better and reasonable behaviour, isn't?)
> 
>     2. have you tried with time accel? (...I'm asking because #4588 have their own limits...)
> 
>     3. Could you try doing the same but with (...or "in") a gas giant?

1. Stars never had this effect in Pioneer, but star heat could solve this problem. Gas giants already do this however; you explode before you can collide with their surface.

2. Yes, same outcome with stars at any time accel. At 10x time accel, same outcome with terrestrial planet terrains (you can pass through the terrain to a certain degree), more than 10x time accel results in loss of ship control -> crashing head-first into the ground -> instant death. This is more of a problem of manual control though.

But, like I said, #4588 seems to have fixed it mostly. Mostly. 

3. See above. Gas giants' atmospheric pressure kills you before you can touch their 'surface'.

Sorry for the late-ish reply.

**Comment by fluffyfreak** [comment]
There won't be any heat/pressure or anything else affecting the ship because it doesn't know that you're near that star, it thinks you're around another "body" in the system, in this case a dummy point we create called a grav-point that has no physical body and is just used to parent binary stars.

When you finally do hit the star it's because the engine has realised that you're closer too another body and switches your frame from the grav-point too the stars.

That's the problem that needs fixing.

**Comment by jaj22** [comment]
@fluffyfreak The key section there is probably lines 634 and 635 of Space.cpp. I don't recognise the code there, but I imagine it's overcompensating for another potential bug with contact binaries. The minimum frame radius there is 1.1 * sbody radius, which is far too small for most practical purposes. It'd hit the minimum if the orbital radius of star A is small (eg. if it's far heavier than star B), or if the calculation is bugged (I haven't checked).

**Comment by fluffyfreak** [comment]
@jaj22 I found the issue that references when that fix was applied for bug #2344

Is there a reason that stars need a frame of reference that is 10 times their radius?

**Comment by fluffyfreak** [comment]
Actually I don't think that's relevent because it gets overridden just after, though I am still curious.

So, in this case it's staying in the parent grav-point's frame for far too long until it eventually breaches the threshold to `SwitchToFrame` in [body.cpp line 233](https://github.com/pioneerspacesim/pioneer/blob/master/src/Body.cpp#L233) which suggests that it's radius is wrong? Because surely by the time you're beneath the surface of a world you should also definitely be within it's radius...

...unless it's radius is wrong because it's an ellipsoid, and "_Ethandce A_" is quite an extreme ellipsoid at that.

Something to look into tomorrow perhaps.

**Comment by fluffyfreak** [comment]
Can anyone chip in and help me understand the `GetRadius` method from [systembody.h](https://github.com/pioneerspacesim/pioneer/blob/master/src/galaxy/SystemBody.h#L104) because it seems to me that by dividing the radius by the aspect ratio (_~1.6 for "Ethandce A"_) you end up with a smaller radius than we should be getting.

@johnbartholomew it seems you wrote that, don't know if you're around much

**Comment by fluffyfreak** [comment]
The commit was in 65fbb93875d0631c4ae35b37cc5ae883e9c67875

**Comment by johnbartholomew** [comment]
For context this was in #2178, which itself was a rebase/cleanup of #1459.

My first guess whenever I see code like this (with my name on it) is that I got the maths wrong from the start. Anyway, I think I can offer a rough explanation of what the code in 65fbb93875d0631c4ae35b37cc5ae883e9c67875 is trying to do, but I don't know whether it makes sense in the wider context of other code that uses these values.

(It looks like) To display an elliptical star body we create a sphere with radius equal to the star's polar radius, and then push out the equator as part of the terrain (TerrainHeightEllipsoid). The aspect ratio adjustment in SystemBody->GetRadius is converting from equatorial to polar radius, and that division still looks correct to me *if* the underlying `radius` field really is equatorial radius (why didn't I clean up the code to only ever store polar radius and just compute equatorial radius when it's needed!? seems dumb). Then the ellipsoid terrain works out its max height from the surface of the (polar-radius) sphere by taking (aspect ratio - 1), which again seems correct assuming max height is supposed to be measured in distance-from-the-surface not distance-from-the-centre.

Does that help, or does it just add confusion?

If I had to take a wild guess I'd say that the equatorial->polar conversion was the right thing to do to get the terrain code working, but that other things (eg, minimum frame radius mentioned in jaj22's post?) we do want equatorial radius, and so trying to use a single "GetRadius" method for both is fundamentally unworkable.
